openstreetmap-carto.lua

```diff
@@ -0,0 +1,518 @@
+    for k, v in pairs(tags) do
+        if roads_info[k] and roads_info[k][v] and roads_info[k][v].roads then
+            if not (k ~= 'railway' or tags.service) then
+                return 1
+            elseif not excluded_railway_service[tags.service] then
+                return 1
+            end
+        end
+    end
+    return 0
+end
+
+-- Filtering on nodes, ways, and relations
+function filter_tags_generic(tags, n)
+    -- Delete tags listed in delete_tags
+    for tag, _ in pairs (tags) do
+        for _, d in ipairs(delete_tags) do
+            if tag == d then
+              tags[tag] = nil
+              break -- Skip this tag from further checks since it's deleted
+            end
+        end
+    end
+    -- By using a second loop for wildcards we avoid checking already deleted tags
+    for tag, _ in pairs (tags) do
+        for _, d in ipairs(delete_wildcards) do
+            if string.find(tag, d) then
+                tags[tag] = nil
+                break
+            end
+        end
+    end
+
+   -- Filter out objects that have no tags after deleting
+    if next(tags) == nil then
+        return 1, {}
+    end
+
+-- Convert layer to an integer
+    tags['layer'] = layer(tags['layer'])
+    return 0, tags
+end
+
+-- Filtering on nodes
+function filter_tags_node (keyvalues, numberofkeys)
+    return filter_tags_generic(keyvalues, numberofkeys)
+end
+
+-- Filtering on relations
+function filter_basic_tags_rel (keyvalues, numberofkeys)
+    -- Filter out objects that are filtered out by filter_tags_generic
+    local filter, keyvalues = filter_tags_generic(keyvalues, numberofkeys)
+    if filter == 1 then
+        return 1, keyvalues
+    end
+
+    -- Filter out all relations except route, multipolygon and boundary relations
+    if ((keyvalues["type"] ~= "route") and (keyvalues["type"] ~= "multipolygon") and (keyvalues["type"] ~= "boundary")) then
+        return 1, keyvalues
+    end
+
+    return 0, keyvalues
+end
+
+-- Filtering on ways
+function filter_tags_way (keyvalues, numberofkeys)
+    local filter = 0  -- Will object be filtered out?
+    local polygon = 0 -- Will object be treated as polygon?
+
+    -- Filter out objects that are filtered out by filter_tags_generic
+    filter, keyvalues = filter_tags_generic(keyvalues, numberofkeys)
+    if filter == 1 then
+        return filter, keyvalues, polygon, roads
+    end
+
+    polygon = isarea(keyvalues)
+
+    -- Add z_order column
+    keyvalues.z_order = z_order(keyvalues)
+
+    return filter, keyvalues, polygon, roads(keyvalues)
+end
+
+--- Handling for relation members and multipolygon generation
+-- Multipolygons are either old-style or new-style. As defined here,
+-- a new-style MP is one with any tags other than type on the relation, and for
+-- them the tags of the ways do not matter.
+--
+-- An old-style MP is one where the only tag on the relation is
+-- type=multipolygon and the tags of all the members are either the same or
+-- empty.
```

Comment:
That's a very strict definition as normally old-style MPs are defined by their outer ways. It would probably be safe to exclude ways with role `inner` without violating the properties below. If it is worth it, is a completely different question.